### ISSUE TITLE
CPU float unary ops can silently return low-precision results: MKL VML first-call kernel dispatch race

### ISSUE THREAD

**Issue by keunhong** [pytorch/pytorch#188792]
### Bug

ATen routes contiguous float unary ops (`sqrt`, `asin`, `exp`, `log`, `sin`, ...) to MKL VML, calling e.g. `vmsSqrt(n, in, out, VML_HA | ...)` from each intra-op worker thread on its chunk (`aten/src/ATen/cpu/vml.h`). oneMKL resolves each VML function's (ISA, accuracy) kernel dispatch lazily on that function's first call. If that first call runs concurrently on several worker threads -- a fresh process's first sizable op, e.g. a dataloader worker's first batch under CPU oversubscription -- one thread can execute a low-accuracy kernel for its whole chunk of that single call. The next call dispatches correctly, so identical reruns are clean and the corruption looks like a transient heisenbug.

Observed bit-exactly on oneMKL 2024.2 (Build 20240605, as bundled in the 2.10.0 wheels) on a Xeon Platinum 8481C: the corrupted chunk of `torch.sqrt` matches the AVX2 EP kernel (`MKL_ENABLE_INSTRUCTIONS=AVX2` + `vmlSetMode(VML_EP)` reproduces it bit-for-bit) despite `VML_HA` being requested per call and AVX512 being the resolved ISA. Errors reach ~4000 ulp (median ~1000). Downstream, sqrt errors of this size can turn `asin(y/r)` with `|y/r|` near 1 into NaN.

### Repro

Waves of spawned processes, each bitwise-comparing its first-ever `torch.sqrt` on 393216 floats against an immediate rerun in the same process (script below). On a 208-core box:

- official 2.10.0+cu128 wheel: ~1 corrupted first-call per 1,500 worker processes
- from-source CPU-only build of current main with statically linked oneMKL 2024.2: 216 / 4,160 (~1 per 19)

The corrupted region is always one `parallel_for` chunk (numel / num_threads), consistent with one worker thread's `vmsSqrt` call running the wrong kernel. Passing `--warmup` as the third argument makes each worker do a serial 1-element `torch.sqrt` first; this drops the rate to 0 (5,200 first-calls), which is the basis of the fix proposed in #188791.

<details>
<summary>mkl_vml_race_repro.py</summary>

```python
"""Minimal repro: MKL VML first-call dispatch race in torch.sqrt.

Spawns waves of worker processes on an oversubscribed box. Each worker makes
its first-ever call to torch.sqrt (which ATen routes to MKL vmsSqrt with
VML_HA from each OpenMP pool thread) and bitwise-compares it against an
immediate rerun on the same tensor. A mismatch means one pool thread's chunk
was computed by a different (low-accuracy EP) kernel on the first call.

Usage: python3 mkl_vml_race_repro.py [waves] [workers] [--warmup]
"""

import multiprocessing
import sys


def worker(worker_id: int, queue, warmup: bool = False) -> None:
    import torch

    torch.set_num_threads(4)
    g = torch.Generator().manual_seed(worker_id)
    s = torch.rand(393216, generator=g) * 2.0 + 1.0
    if warmup:
        # Emulates the proposed vml.h fix: a serial 1-element VML call
        # (below the parallel_for grain) resolves MKL dispatch first.
        torch.sqrt(torch.tensor([0.5]))
    r1 = torch.sqrt(s)  # first parallel VML call in this process
    r2 = torch.sqrt(s)
    if torch.equal(r1, r2):
        queue.put(None)
        return
    diff = r1.view(torch.int32) != r2.view(torch.int32)
    idx = diff.nonzero().squeeze(-1)
    ulp = (r1.view(torch.int32).long() - r2.view(torch.int32).long()).abs()[diff]
    queue.put(
        {
            "worker": worker_id,
            "diff_elems": int(diff.sum()),
            "span": (int(idx[0]), int(idx[-1])),
            "numel": s.numel(),
            "max_ulp": int(ulp.max()),
            "median_ulp": int(ulp.median()),
        }
    )


def main() -> int:
    waves = int(sys.argv[1]) if len(sys.argv) > 1 else 10
    workers = int(sys.argv[2]) if len(sys.argv) > 2 else 104
    warmup = len(sys.argv) > 3 and sys.argv[3] == "--warmup"
    ctx = multiprocessing.get_context("spawn")
    total = 0
    hits = []
    for wave in range(waves):
        queue = ctx.SimpleQueue()
        procs = [
            ctx.Process(
                target=worker, args=(wave * workers + i, queue, warmup), daemon=True
            )
            for i in range(workers)
        ]
        for p in procs:
            p.start()
        for p in procs:
            p.join()
        while not queue.empty():
            result = queue.get()
            total += 1
            if result is not None:
                hits.append(result)
                print(f"HIT wave={wave}: {result}", flush=True)
        print(f"wave {wave} done: {total} workers so far, {len(hits)} hits", flush=True)
    print(f"TOTAL: {total} first-calls, {len(hits)} bitwise mismatches")
    return 0 if hits else 1


if __name__ == "__main__":
    raise SystemExit(main())
```

</details>

Example hit (unpatched, from-source static-MKL build; span is one quarter of the tensor = one of 4 worker threads' chunks):

```
HIT wave=0: {'worker': 37, 'diff_elems': 98265, 'span': (294912, 393215), 'numel': 393216, 'max_ulp': 3935, 'median_ulp': 1040}
```

### oneMKL version matrix

Same storm repro (40 waves x 104 workers = 4,160 first-calls per cell); unpatched from-source builds differing only in the statically linked oneMKL:

| oneMKL (static) | corrupted first-calls / 4,160 |
|---|---|
| 2024.2.0 (current CI/wheel pin) | 216 |
| 2024.2.1 | 295 |
| 2024.2.2 | 224 |
| 2025.0.0 | 0 |
| 2025.2.0 | 0 |
| 2026.1.0 | 0 |

So the underlying race appears fixed in oneMKL 2025.0. CI and wheels pin `mkl-static==2024.2.0` (`.ci/docker/requirements-ci.txt`, `.ci/docker/common/install_mkl.sh`), so shipped binaries remain affected. Bumping the pin is an alternative or complementary fix to #188791.

### Versions

- torch 2.10.0+cu128 wheel (bundles oneMKL 2024.2 Product Build 20240605), and current main (`1a0a3c99f32`) built from source CPU-only with `BLAS=MKL`, statically linked oneMKL 2024.2
- Intel Xeon Platinum 8481C (Sapphire Rapids, AVX512), 208 threads, Linux 6.17
- Any op routed through `IMPLEMENT_VML_MKL` in `aten/src/ATen/cpu/vml.h` is exposed (sqrt, asin, acos, atan, cos, sin, tan, tanh, exp, log, log10, log2, erf, erfc, erfinv, trunc); sqrt is just the easiest to catch

Fix proposed in #188791: a per-function serialized 1-element warm-up call under `c10::call_once` so MKL resolves its dispatch before any concurrent use. The underlying race is inside closed-source oneMKL and is being reported to Intel separately.

cc @ezyang @gchanan @kadeng @msaroufim @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01 @albanD @pragupta @frank-wei

**Comment by albanD** [comment]
High priority to investigate
